text/1004-transactions-api.rst

```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
+   specified number of seconds until connection is established (at least
+   single one for connection pool).
+2. If first connection could not be established for this timeout, error
+   is thrown that holds the reason of the last failure.
+3. On the subsequent reconnects this timeout is also obeyed (i.e.
+   connection might try to reconnect multiple times) and error is thrown
+   after the specified timeout as a result of the operation that
+   requires connection (i.e. on query or transaction start).
+4. For ``retry`` we bail out if this timeout is reached during any
+   single reconnect. We don't retry immediately after database is marked
+   as unavailable.
+5. Subsequent queries or transactions after failure will retry for the
+   specified timeout each time.
+
+Only the following conditions are treated as eligible for reconnect:
+
+1. Name resolution failed
+2. File not found (Unix socket not bound yet)
+3. Connection reset, connection aborted, connection refused (server is
+   restarting or not ready yet)
+4. Timeout happened during connect or authentication
+
+All other errors are propagated immediately.
+
+This timeout is different from ``connect_timeout``. Connect timeout
+determines how long individual connect attempt may take. And
+``wait_until_available`` determines how many such attempts could be
+made (i.e. how many ones fits the time frame).
+
+
+EdgeDB Changes
+--------------
+
+To support features above we add two headers to EdgeDB queries:
 
-While this covers "concurent update" and "deadlock" errors, with the
-current implementation it can't reconnect when when connection was
-interrupted while transaction is active. There are few ways it can work:
+1. For PrepareComplete_, CommandComplete_ server-side messages:
+   ``SERVER_OPT_HAS_FEATURES``
+2. For Prepare_, OptimisticExecute_, Execute_, ExecuteScript_
+   client-side messages: ``QUERY_OPT_ALLOW_FEATURES``
 
-a. It may only retry on the same connection and fail on disconnect
-b. It may reconnect and replace underlying socket in the connection
-   object and retry
-c. We may only allow transactions on connection pools (current
-   specification)
+Both contain the set of strings joined by comma:
 
-This can be postponed to later RFC when we know what are use cases of
-it.
+1. `modification` -- query is not read-only
+2. `session` -- query contains session config change
+3. `config` -- server or database config change
+4. `ddl` -- query contains DDL
+5. `transaction` -- query contains start/commit/rollback of transaction
+    or savepoint manipulation
 
+In case of ``SERVER_OPT_HAS_FEATURES`` it describes what is actually
+contained in the query. And in case of ``QUERY_OPT_ALLOW_FEATURES``
+client can specify what of these things are allowed in this query.
 
-Retry Single Queries
---------------------
+Read-only queries are always allowed. When ``QUERY_OPT_ALLOW_FEATURES``
+is omitted any query is allowed (default). With the empty string only
+read-only queries are allowed.
```

Comment:
I don't understand why we need this complexity.  IMO instead of "modification" header flag we should have "readonly". If it's set the query fails to prepare unless it's readonly. This is way easier to understand and document.

Reply:
This is because:
```
conn.execute("""
START TRANSACTION;
UPDATE User SET {...}
COMMIT
""")
```
Would return `transaction,modifications` flag.

With @1st1 proposal `.execute("START TRANSACTION")` would return `transaction,readonly` features and `.execute("START TRANSACTION; UPDATE..."` would return `transaction`.

I.e. purely additive features are better at least for `HAS_FEATURES`, but is not bad for `ALLOW` too.

I we go with bitmask read-only queries == '0' is also good enough.

Reply:
I agree with @tailhook.  It's easier to reason when each bit _extends_ capabilities consistently as opposed to either restricting or extending them.